### PR TITLE
Support declaration and consumption of user-defined Unsigned Right Shift operator.

### PR DESCRIPTION
**Pull request by AlekseyTs**
https://github.com/dotnet/csharplang/blob/main/proposals/unsigned-right-shift-operator.md

Also added language version checks.

Test plan https://github.com/dotnet/roslyn/issues/60433

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@jcouv, @333fred, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@jcouv, @333fred, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@jcouv, @333fred, @dotnet/roslyn-compiler Please review.

**Comment by jcouv** [comment]
Looking

File: src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs [hunk 1]
```diff
@@ -1869,5 +1937,1583 @@ static void Main()
...
+}
+
+public class C4
+{
+    public static int operator >>>(C4 x, int y)
+    {
+        throw null;
+    }
+
+    public static int operator >>(C4 x, int y)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs:2125]: Consider adding a positive case for parameters being nullable: `public static int operator >>>(S1? x, int? y)`

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 2]
```diff
@@ -2342,10 +2348,28 @@ private bool CheckConstraintLanguageVersionAndRuntimeSupportForOperator(SyntaxNo
...
+                {
+                    result &= CheckFeatureAvailability(node, MessageID.IDS_FeatureUnsignedRightShift, diagnostics);
+                }
+            }
+            else
             {
-                result &= CheckFeatureAvailability(node, MessageID.IDS_FeatureCheckedUserDefinedOperators, diagnostics);
+                Debug.Assert((methodOpt.Name == WellKnownMemberNames.UnsignedRightShiftOperatorName) == isUnsignedRightShift);
+
+                if (Compilation.SourceModule != methodOpt.ContainingModule)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:2362]: I didn't understand why the unsigned right shift operator LangVer check is conditioned on resolved method being in source. I would have expected an unconditional LangVer diagnostic for `>>>` where ever the best signature is found.

File: src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs [hunk 3]
```diff
@@ -1037,6 +1037,8 @@ CurrentToken.Kind is (SyntaxKind.GreaterThanToken or SyntaxKind.GreaterThanEqual
                                 operatorToken.Text + operatorToken2.Text + operatorToken3.Text,
                                 operatorToken.ValueText + operatorToken2.ValueText + operatorToken3.ValueText,
                                 operatorToken3.GetTrailingTrivia());
+
+                            operatorToken = CheckFeatureAvailability(operatorToken, MessageID.IDS_FeatureUnsignedRightShift, forceWarning: true);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs:1041]: nit: Why do a parser LangVer check as opposed to add check in Binder_Cref?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceUserDefinedOperatorSymbolBase.cs [hunk 4]
```diff
@@ -358,6 +358,7 @@ private void CheckOperatorSignatures(BindingDiagnosticBag diagnostics)
 
                 case WellKnownMemberNames.LeftShiftOperatorName:
                 case WellKnownMemberNames.RightShiftOperatorName:
+                case WellKnownMemberNames.UnsignedRightShiftOperatorName:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceUserDefinedOperatorSymbolBase.cs:361]: nit: consider updating comment in `CheckShiftSignature` (add `>>>`)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:2362]: >I didn't understand why the unsigned right shift operator LangVer check is conditioned on resolved method being in source. I would have expected an unconditional LangVer diagnostic for ```>>>``` where ever the best signature is found.

1. This logic is not new
2. If the operator is declared in the same module (note, not just in source), we perform the check on the declaration and report the error, if needed. Reporting the same error at every consumption site would only add more noise.

File: src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs:1041]: > Why do a parser LangVer check as opposed to add check in Binder_Cref?

Just following the established pattern.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs:1041]: Note that these are warnings, not errors

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler For the second review.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

File: src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs [hunk 5]
```diff
@@ -1869,5 +1937,1650 @@ static void Main()
...
+
+        if (Test1(null, 1) is null) System.Console.WriteLine(""Passed 2"");
+
+        if (Test1(new C1(), null) is null) System.Console.WriteLine(""Passed 3"");
+
+        if (Test1(null, null) is null) System.Console.WriteLine(""Passed 4"");
+    }
+
+    static C1? Test1(C1? x, int? y) => x >>> y; 
+    static C1? Test2(C1? x, int? y) => x >> y;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs:2463]: Unused?

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 3)

File: src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs [hunk 6]
```diff
@@ -1869,5 +1937,1650 @@ static void Main()
...
+    static C1 Test1(C1 x, int y) => C1.op_UnsignedRightShift(x, y); 
+}
+";
+
+            var compilation1 = CreateCompilationWithIL(source1, ilSource, options: TestOptions.DebugExe,
+                                                       parseOptions: TestOptions.RegularPreview);
+            // PROTOTYPE(UnsignedRightShift): This code was previously allowed. Confirm that we are Ok with this 
+            //                                breaking change and document it.  
+            compilation1.VerifyDiagnostics(
+                // (9,40): error CS0571: 'C1.operator >>>(C1, int)': cannot explicitly call operator or accessor
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs:2069]: > Would it be worth adding a version of this test without the specialname flag in the IL?

A method like that can be defined in plain C#.

File: src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs:2463]: > Unused?

I think it is used for IL comparison

File: src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs [hunk 7]
```diff
@@ -1869,5 +1937,1650 @@ static void Main()
...
+        if (Test1(null, null) is null) System.Console.WriteLine(""Passed 4"");
+    }
+
+    static C1? Test1(C1? x, int? y)
+    {
+        x >>>= y;
+        return x;
+    }
+
+    static C1? Test2(C1? x, int? y)
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/UnsignedRightShiftTests.cs:2557]: > Unused?

Used for IL comparison

**Review by 333fred — APPROVED** [review]

**Pull request merged**